x/gov/model.go

```diff
@@ -95,22 +105,80 @@ func (m Fraction) Validate() error {
 		return errors.Wrap(errors.ErrInvalidInput, "numerator must not be 0")
 	case m.Denominator == 0:
 		return errors.Wrap(errors.ErrInvalidInput, "denominator must not be 0")
-	case m.Numerator*2 < m.Denominator:
+	case uint64(m.Numerator)*2 < uint64(m.Denominator):
 		return errors.Wrap(errors.ErrInvalidInput, "must not be lower 0.5")
-	case m.Numerator/m.Denominator > 1:
+	case m.Numerator > m.Denominator:
 		return errors.Wrap(errors.ErrInvalidInput, "must not be greater 1")
 	}
 	return nil
 }
 
+const (
+	minDescriptionLength = 3
+	maxDescriptionLength = 5000
+)
+
 func (m *TextProposal) Validate() error {
-	// TODO impl
+	switch {
+	case !validTitle(m.Title):
+		return errors.Wrap(errors.ErrInvalidInput, fmt.Sprintf("title: %q", m.Title))
+	case m.Status == TextProposal_Invalid:
+		return errors.Wrap(errors.ErrInvalidInput, "invalid status")
+	case !m.VotingStartTime.Time().Before(m.VotingEndTime.Time()):
```

Comment:
`weave.UnixTime` is just a number, you can as well write `case m.VotingStartTime >= m.VotingEndTime:`.